File: leetcode/439-Ternary-Expression-Parser/Solution.py
class Solution:
    def parseTenary(self, expression: str) -> str:
        stack = []

        for char in str[::-1]:
            if stack and stack[-1] == '?':
                stack.pop()
                onTrue = stack.pop()
                stack.pop()
                onFalse = stack.pop()
                stack.append(onTrue if char == 'T' else onFalse)
            else:
                stack.append(char)
        return stack[0]

File: leetcode/439-Ternary-Expression-Parser/Solution.kt
class Solution {
    fun parseTernary(expression: String): String {
        val stack = ArrayDeque<Char>()
        val i = expression.length - 1

        while (i >= 0) {
            val char = expression[i]
            when {
                char == '?' -> {
                    val onTrue = stack.removeLast()
                    val onFalse = stack.removeLast()
                    val condition = expression[--i]
                    stack.addLast(if (condition == 'T') onTrue else onFalse)
                }
                char.isDigit() || char == 'T' || char == 'F' -> {
                    stack.addLast(char)
                }
            }
            i--
        }
        return stack.removeLast().toString()
    }
}